gowin/main.cc

```diff
@@ -54,20 +54,21 @@ po::options_description GowinCommandHandler::getArchOptions()
 
 std::unique_ptr<Context> GowinCommandHandler::createContext(dict<std::string, Property> &values)
 {
-    std::regex devicere = std::regex("GW1N([A-Z]*)-(LV|UV)([0-9])([A-Z]{2}[0-9]+)(C[0-9]/I[0-9])");
+    std::regex devicere = std::regex("GW1N([A-Z]*)-(LV|UV|UX)([0-9])(C?)([A-Z]{2}[0-9]+)(C[0-9]/I[0-9])");
     std::smatch match;
     std::string device = vm["device"].as<std::string>();
     if (!std::regex_match(device, match, devicere)) {
         log_error("Invalid device %s\n", device.c_str());
     }
     ArchArgs chipArgs;
     char buf[36];
-    snprintf(buf, 36, "GW1N%s-%s", match[1].str().c_str(), match[3].str().c_str());
+    snprintf(buf, 36, "GW1N%s-%s%s", match[1].str().c_str(), match[3].str().c_str(),
+		match[4].str().c_str());
     chipArgs.device = buf;
-    snprintf(buf, 36, "GW1N-%s", match[3].str().c_str());
+    snprintf(buf, 36, "GW1N%s-%s", match[1].str().c_str(), match[3].str().c_str());
```

Comment:
This breaks GW1NR-9 actually because it's the same DB as GW1N-9, but now it tries to load the non-existent GW1NR9 DB.

At this point it's unknown if GW1NS-2 is actually different from GW1N-2, because nobody bothered to add support for it yet. So... not sure if we should special-case "S" or just revert this line.